### PR TITLE
Fix NoDatabaseError when creating a new tenant

### PR DESCRIPTION
**Pull request by tachyons**
We are exploring this fork as we migrate our app to Rails6, this is the first blocker we encountered

### CODE REVIEW

**Comment by rpbaltazar** [comment]
In which scenario does it break for you? We are using this in rails 6 in production (and development), using postgis adapter and we haven't encountered any issues. can you tell us a bit more about the setup that you have, so we can try to reproduce it in tests

**Comment by tachyons** [comment]
Sure, I will make a reproduction repo, The issue is that we use a static list of tenants, like
```
config.tenants = ["first_tenant", "second_tenant", "third_tenant"]
```

So looping through these tenants before database is being created will cause No database error

**Comment by ahorek** [comment]
Hi, I have a very similar problem related to this PR. I can't create a new database, because the database doesn't exist yet :)

it fails on NoDatabase errors and also on ActiveRecord::StatementInvalid here:
https://github.com/rails-on-services/apartment/blob/master/lib/apartment/adapters/mysql2_adapter.rb#L42

rake db:create
```
Mysql2::Error: Unknown database 'public'
/lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:131:in `_query'
/lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:131:in `block in query'
/lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:130:in `handle_interrupt'
/lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:130:in `query'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:187:in `block (2 levels) in execute'
/lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:186:in `block in execute'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:581:in `block (2 levels) in log'
/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `block in log'
/lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:571:in `log'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:185:in `execute'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/mysql/database_statements.rb:28:in `execute'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:42:in `reset'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:36:in `initialize'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:9:in `new'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:9:in `mysql2_adapter'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/tenant.rb:56:in `adapter'
/lib/ruby/2.4.0/forwardable.rb:223:in `init'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/tenant.rb:25:in `init_once'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:39:in `block in connection'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:38:in `tap'
/lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:38:in `connection'
lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/mysql_database_tasks.rb:6:in `connection'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/mysql_database_tasks.rb:14:in `create'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:119:in `create'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:139:in `block in create_current'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:316:in `block in each_current_configuration'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:313:in `each'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:313:in `each_current_configuration'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:138:in `create_current'
/lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/railties/databases.rake:29:in `block (2 levels) in <top (required)>'
/lib/ruby/gems/2.4.0/gems/rake-13.0.1/lib/rake/task.rb:281:in `block in execute'
```

note that https://github.com/influitive/apartment 2.2.1 does work for me.

see also https://github.com/influitive/apartment/pull/546

File: lib/apartment/railtie.rb [hunk 1]
```diff
@@ -39,11 +39,13 @@ def connection
         super.tap do
           Apartment::Tenant.init_once
```
Comment (ahorek) [inline lib/apartment/railtie.rb:40]: I think ```super``` shouldn't be guarded by ActiveRecord::NoDatabaseError.

it breaks my case where I expect the connection will raise ActiveRecord::NoDatabaseError, but now it returns nil. The same for ```arel_table```

**Comment by rpbaltazar** [comment]
> Hi, I have a very similar problem related to this PR. I can't create a new database, because the database doesn't exist yet :)
> 
> it fails on NoDatabase errors and also on ActiveRecord::StatementInvalid here:
> https://github.com/rails-on-services/apartment/blob/master/lib/apartment/adapters/mysql2_adapter.rb#L42
> 
> rake db:create
> 
> ```
> Mysql2::Error: Unknown database 'public'
> /lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:131:in `_query'
> /lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:131:in `block in query'
> /lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:130:in `handle_interrupt'
> /lib/ruby/gems/2.4.0/gems/mysql2-0.5.3-x64-mingw32/lib/mysql2/client.rb:130:in `query'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:187:in `block (2 levels) in execute'
> /lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
> /lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
> /lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:186:in `block in execute'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:581:in `block (2 levels) in log'
> /lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `block in log'
> /lib/ruby/gems/2.4.0/gems/activesupport-5.2.4.2/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_adapter.rb:571:in `log'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:185:in `execute'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/mysql/database_statements.rb:28:in `execute'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:42:in `reset'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:36:in `initialize'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:9:in `new'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/adapters/mysql2_adapter.rb:9:in `mysql2_adapter'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/tenant.rb:56:in `adapter'
> /lib/ruby/2.4.0/forwardable.rb:223:in `init'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/tenant.rb:25:in `init_once'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:39:in `block in connection'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:38:in `tap'
> /lib/ruby/gems/2.4.0/gems/ros-apartment-2.6.1/lib/apartment/railtie.rb:38:in `connection'
> lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/mysql_database_tasks.rb:6:in `connection'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/mysql_database_tasks.rb:14:in `create'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:119:in `create'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:139:in `block in create_current'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:316:in `block in each_current_configuration'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:313:in `each'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:313:in `each_current_configuration'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/tasks/database_tasks.rb:138:in `create_current'
> /lib/ruby/gems/2.4.0/gems/activerecord-5.2.4.2/lib/active_record/railties/databases.rake:29:in `block (2 levels) in <top (required)>'
> /lib/ruby/gems/2.4.0/gems/rake-13.0.1/lib/rake/task.rb:281:in `block in execute'
> ```
> 
> note that https://github.com/influitive/apartment 2.2.1 does work for me.
> 
> see also [influitive#546](https://github.com/influitive/apartment/pull/546) and #67

I think yours is related to this
https://github.com/rails-on-services/apartment/issues/61

**Comment by rpbaltazar** [comment]
> Sure, I will make a reproduction repo, The issue is that we use a static list of tenants, like
> 
> ```
> config.tenants = ["first_tenant", "second_tenant", "third_tenant"]
> ```
> 
> So looping through these tenants before database is being created will cause No database error

I actually have a [demo app with postgres](https://github.com/rails-on-services/rails-6-demo/tree/running-master-branch-ros-apartment) that seems to work. Are you using postgres?

**Comment by ahorek** [comment]
I'm patching a relation before my db is created, which is a stupid thing I know. However, it would be nice to address this failure. It does work without the apartment gem.
```
ActiveRecord::NoDatabaseError: Unknown database '22fcb929c87b7cbfbd056b8016a33015'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/mysql2_adapter.rb:26:in `rescue in mysql2_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/mysql2_adapter.rb:12:in `mysql2_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:830:in `new_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:874:in `checkout_new_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:853:in `try_to_checkout_new_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:814:in `acquire_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:538:in `checkout'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:382:in `connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:1033:in `retrieve_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_handling.rb:118:in `retrieve_connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/connection_handling.rb:90:in `connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/railtie.rb:39:in `connection'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/adapters/mysql2_adapter.rb:44:in `reset'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/adapters/mysql2_adapter.rb:36:in `initialize'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/adapters/mysql2_adapter.rb:9:in `new'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/adapters/mysql2_adapter.rb:9:in `mysql2_adapter'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/tenant.rb:56:in `adapter'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/tenant.rb:25:in `init_once'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/bundler/gems/apartment-8eeed0e2dd12/lib/apartment/railtie.rb:45:in `arel_table'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/relation.rb:25:in `initialize'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/relation/delegation.rb:132:in `new'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/relation/delegation.rb:132:in `create'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/core.rb:285:in `relation'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/scoping/named.rb:50:in `default_scoped'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/scoping/named.rb:36:in `all'
/usr/local/rvm/gems/ruby-2.6.3-railsexpress/gems/activerecord-5.2.4.2/lib/active_record/querying.rb:12:in `preload'
```

**Comment by rpbaltazar** [comment]
@tachyons can you try the latest gem version (2.7.0) and let me know if that fixed the problem? If not, can you please open an issue.

**Comment by rpbaltazar** [comment]
> I'm patching a relation before my db is created, which is a stupid thing I know. However, it would be nice to address this failure. It does work without the apartment gem.

Can you open an issue with a sample of what you're doing? I'm not sure what are you trying to achieve that works without apartment but apartment breaks...

**Comment by rpbaltazar** [comment]
Closing this for now.

**Comment by tachyons** [comment]
@rpbaltazar Here is a simple reproduction repo, try running `bundle exec rake db:create` on this repo

**Pull request closed without merging**